[MRG] Jail

aapeliv opened the pull request:
As discussed on Slack, I've implemented a prototype "jail" idea for review/comments. I'm not sure if this is a good way of solving these problems, so I'd like some feedback on what people think. It feels kind of convoluted and especially tacking it on now to the authentication system (that assumed a 2-state model) makes it a bit awkward (there's now 3 states: not logged in, logged in, and logged in but jailed).

This makes it possible for a user to be authenticated while not being able to use the main APIs, but only a subset, the "jailed" ones. This is useful if:

1. The user has not finished signing up (e.g. has not finished adding a location)
2. The user has not accepted TOS/GDPR/etc
3. The user is banned, but we'd like to show them something nicer than a "you're banned" at login

It makes it easy to add requirements, e.g. update a privacy policy and make sure everyone has agreed to them, or add required fields to a user profile.

This is also missing tests.

aapeliv committed:
Pass full session to auth interceptor

aapeliv committed:
Implement jail idea

aapeliv committed:
Add allow_jailed=False to auth interceptors in test

kalvdans commented on app/backend/src/couchers/models.py:
```diff
@@ -176,6 +182,17 @@ def mutual_friends(self, target_id):
+        """
+        Sync all sessions with current jail status.
+
+        The auth interceptor does not check users table to reduce per-request DB overhead and this info is cached in the user session
```
Add more indices to the database instead if you worry about performance. No duplicated state please!

kalvdans commented on app/backend/src/couchers/interceptors.py:
```diff
@@ -36,16 +37,19 @@ def intercept_service(self, continuation, handler_call_details):
             return unauthenticated_handler()
 
+        if not self._allow_jailed and user_session.jailed:
+            return unauthenticated_handler("Jailed")
```
Maybe don't leak the word Jail to the end user. (some clients may show the error reason straight up)

aapeliv replied:
We should pick some appropriate word.

kalvdans reviewed (changes requested):
Please versionify TOS (int instead of bool) to easier allow new versions of TOS to be deployed without updating the database.

aapeliv committed:
Change "Jailed" to "Permission denied" in (possibly) user-facing error message

aapeliv committed:
Change TOS to integer

Default to 0 or 1 for now, leave extension to larger version until later

aapeliv committed:
Fix is_jailed logic

aapeliv committed:
Don't duplicate jail logic in UserSession

Comes at the expense of an extra join in each auth check

aapeliv commented:
I've changed the TOS field to an integer, and removed the duplicated state. I haven't added any logic to actually version the TOS field, I'll leave that to later. I wasn't actually thinking yet as far as to put some real TOS, it was just a "sample placeholder" jail thing. As @NightTsarina suggested, I'm trying to keep PRs smaller, and wanted feedback on jail before slapping the user location on top of it.

kalvdans commented on app/backend/src/couchers/servicers/auth.py:
```diff
@@ -48,19 +48,23 @@ def get_auth_interceptor(self, allow_jailed):
 
-            if not user_session:
+            if not result:
                 # can't expunge if it's None
```
Stale comment, remove.

kalvdans reviewed (changes requested):
Please add a comment in the top of the the unjailed API descriptions files (api.proto, conversations.proto, media.proto, requests.proto). Something along these lines:
```
// All these calls can generate Unauthenticated Error for the following reasons: 
// Session token expired, user blocked, TOS changed or New profile data needed.
// Call `JailInfo()` to get the exact reason.
```

aapeliv committed:
Fix review comments

aapeliv committed:
Fix TOS.vue being in wrong path

aapeliv committed:
Fix most tests

aapeliv committed:
Add jailed parameter to generate_user test fixture

aapeliv committed:
Fix up CompleteTokenLogin with database join instead of lazy-loading user

aapeliv committed:
Start on jail tests

aapeliv committed:
Finish jail tests

aapeliv commented:
I added tests. This is almost ready, except the `TOS.vue` page should probably say something a bit more elaborate than `I understand Couchers.org is not a dating platform, etc, etc.`.

Also, I hope to merge #319 and update migrations before merging this.

aapeliv committed:
Change jail API to be more sensible

aapeliv committed:
Add  field to JailInfoRes for client side simplicity

aapeliv committed:
Update vue frontend to match new jail API

aapeliv commented:
How about "paddock", "cage", or just "box"?

lucaslcode commented:
How about "pending"? Like a pending user.

lucaslcode commented on app/backend/src/couchers/models.py:
```diff
@@ -108,6 +108,12 @@ class User(Base):
+
+    @property
+    def is_jailed(self):
+        return self.accepted_tos < 1
```
Should we have '1' stored in config somewhere as the current TOS version?

aapeliv replied:
Yes, there should be some logic for this, but I didn't want to do it yet as it's not very relevant yet.

Actually, maybe all it takes is that when the TOS are updated, we do an `UPDATE users SET accepted_tos = 0` :P

kalvdans replied:
The frontend needs to pass in which version of the TOS the user has agreed to.

aapeliv replied:
@kalvdans: could we leave the rest of versioned TOS to a later feature?

kalvdans replied:
Sure, at the expense of more database migrations/API  incompatibilities.

lucaslcode commented on app/backend/src/tests/test_jail.py:
```diff
@@ -0,0 +1,147 @@
+
+    with real_jail_session(db, token1) as jail:
+        res = jail.JailInfo(empty_pb2.Empty())
+        # check every field is true
```
false?

aapeliv replied:
Yep, you're right

lucaslcode commented on app/vue/src/views/TOS.vue:
```diff
@@ -0,0 +1,65 @@
+        :loading="loading"
+        :disabled="acceptedTos"
+        v-on:click="acceptTos"
+        >{{ acceptedTos ? "Already accpeted" : "Accept" }}</v-btn
```
```suggestion
        >{{ acceptedTos ? "Already accepted" : "Accept" }}</v-btn
```

aapeliv replied:
Thanks

lucaslcode commented on app/vue/src/views/TOS.vue:
```diff
@@ -0,0 +1,65 @@
+      <error-alert :error="error" />
+      <h1>Terms of Service</h1>
+      <p>Please accept the TOS before continuing to Couchers.org</p>
+      <p>I understand Couchers.org is not a dating platform, etc, etc.</p>
```
```suggestion
      <p>This is a test TOS. I agree to follow the <a href="https://community.couchers.org/faq">Couchers guidelines</a>. I understand this is a preview of couchers.org and my data may be erased.</p>
```

lucaslcode commented on app/backend/src/tests/test_fixtures.py:
```diff
@@ -44,48 +46,82 @@ def db():
+        countries_visited="Testing country",
+        countries_lived="Wonderland",
+        # you need to make sure to update this logic to make sure the user is jailed/not on request
+        accepted_tos=0 if jailed else 1,
```
This could make testing of other jailing reasons difficult. Maybe for now you could change the argument from 'jailed' to 'acceptedTOS'?

aapeliv replied:
Actually, on a second thought, I don't think this should be a parameter here anyway, the jail test modifies the user as needed anyway. I'll refactor it out.

lucaslcode reviewed (dismissed):
Only minor nitpicks

aapeliv committed:
Fix minor review comments

aapeliv committed:
Remove jailed parameter from test user generation

aapeliv commented:
@lucaslcode: does this require some quick patch on the react frontend so signup still works and it's not bricked?

lucaslcode commented:
On it #342

lucaslcode committed:
Add jail compatibility

Add TOS
closes #342

lucaslcode committed:
Review jail frontend

kalvdans reviewed (dismissed):
Thanks for documenting the calling contexts in the API descriptions! LGTM!

lucaslcode committed:
Fix setState after unmount

lucaslcode commented:
@aapeliv I have to resolve these merge frontend conflicts but there's scary alembic errors, what's the best thing to do? Merge develop into this branch and let you fix alembic?

lucaslcode commented:
I made the commit, feel free to revert it

aapeliv commented:
> @aapeliv I have to resolve these merge frontend conflicts but there's scary alembic errors, what's the best thing to do? Merge develop into this branch and let you fix alembic?

Thanks, I'll fix it and merge

aapeliv committed:
Fix migrations

aapeliv committed:
Fix jail in client.ts

aapeliv committed:
Fix up frontend/package.json
